src/roswire/definitions/package.py

```diff
@@ -82,15 +83,28 @@ def to_dict(self) -> Dict[str, Any]:
 
 class PackageDatabase(Mapping[str, Package]):
     @staticmethod
-    def paths(shell: ShellProxy) -> List[str]:
+    def paths(shell: ShellProxy, file_proxy: FileProxy) -> List[str]:
         """
         Parses the contents of the ROS_PACKAGE_PATH environment variable for a
         given shell.
         """
         code, path_str, duration = shell.execute('echo "${ROS_PACKAGE_PATH}"')
         if code != 0:
             raise Exception("unexpected error when fetching ROS_PACKAGE_PATH")
-        paths: List[str] = path_str.strip().split(':')
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            if file_proxy.isfile(f'{path}/package.xml'):
+                paths.append(path)
+            else:
+                try:
+                    dirs: List[str] = file_proxy.listdir(path, absolute=True)
```

Comment:
You could just extend a queue with all of the subdirectories at this point.

Reply:
The comment above.